### PR TITLE
Vector tile 2.x, sort during

### LINKED ISSUE #36: radius=0 leads to zero results with PIP
When radius is `0`, we expect point in polygon style results (ie 1 result if we’re within a single polygon). Right now we get zero results since the query point is a double (lng lat) but by the time we’re comparing the distance from closest_point we’re in integers, which then get converted to a lng lat and can be different than the original query point since it’s an interpolation. This means the distance check looks something like `if (0.00011241241 <= 0) { ... }` which evaluates to false.

Three possible solutions: 
1. We’ll want to use something other than closest_point for PIP (radius=0) operations
2. Do all calculations in doubles instead of ints
3. Instead of using zero (if radius is zero) we use some sort of very small number like `if (0.0000234 < 0.5)` to catch these instances

cc @flippmoke

### PR DESCRIPTION
**Pull request by mapsam**
Integrating vector tile 2.x and a "sort during" 

### What changed?

* using a std::priority_queue for the ResultObject references, which is sorted upon emplacing into the queue, rather than sorting after the entire loop finishes
* integrate vector-tile 2.x branch to decode geometry and properties for us (removes the geometry_processors.cpp file) - includes a new visitor function for decoding variants into v8 values
* adds a separate code path for radius `0.0`, resolves #36 (closing #41) 

### benchmarks

**this branch**

```
UV_THREADPOOL_SIZE=1 node bench/vtquery.bench.js --iterations=1000 --concurrency=1

1: pip: many building polygons ... 956 runs/s (1046ms)
2: pip: many building polygons, single layer ... 1053 runs/s (950ms)
3: query: many building polygons, single layer ... 928 runs/s (1078ms)
4: query: linestrings, mapbox streets roads ... 1709 runs/s (585ms)
5: query: polygons, mapbox streets buildings ... 935 runs/s (1069ms)
6: query: all things - dense single tile ... 510 runs/s (1962ms)
7: query: all things - dense nine tiles ... 65 runs/s (15418ms)
8: elevation: terrain tile nepal ... 849 runs/s (1178ms)
9: geometry: 2000 points in a single tile, no properties ... 2053 runs/s (487ms)
10: geometry: 2000 points in a single tile, with properties ... 1645 runs/s (608ms)
11: geometry: 2000 linestrings in a single tile, no properties ... 1086 runs/s (921ms)
12: geometry: 2000 linestrings in a single tile, with properties ... 952 runs/s (1050ms)
13: geometry: 2000 polygons in a single tile, no properties ... 690 runs/s (1450ms)
14: geometry: 2000 polygons in a single tile, with properties ... 643 runs/s (1556ms)
```

**master**

```
UV_THREADPOOL_SIZE=1 node bench/vtquery.bench.js --iterations=1000 --concurrency=1

1: pip: many building polygons ... 1100 runs/s (909ms)
2: pip: many building polygons, single layer ... 1248 runs/s (801ms)
3: query: many building polygons, single layer ... 356 runs/s (2806ms)
4: query: linestrings, mapbox streets roads ... 685 runs/s (1459ms)
5: query: polygons, mapbox streets buildings ... 362 runs/s (2766ms)
6: query: all things - dense single tile ... 242 runs/s (4127ms)
7: query: all things - dense nine tiles ... 39 runs/s (25482ms)
8: elevation: terrain tile nepal ... 947 runs/s (1056ms)
9: geometry: 2000 points in a single tile, no properties ... 1410 runs/s (709ms)
10: geometry: 2000 points in a single tile, with properties ... 511 runs/s (1956ms)
11: geometry: 2000 linestrings in a single tile, no properties ... 875 runs/s (1143ms)
12: geometry: 2000 linestrings in a single tile, with properties ... 423 runs/s (2366ms)
13: geometry: 2000 polygons in a single tile, no properties ... 715 runs/s (1398ms)
14: geometry: 2000 polygons in a single tile, with properties ... 378 runs/s (2646ms)
```

cc @flippmoke

### CODE REVIEW

File: src/vtquery.cpp [hunk 1]
```diff
@@ -324,12 +342,12 @@ struct Worker : Nan::AsyncWorker {
 
             // create properties object
             v8::Local<v8::Object> properties_obj = Nan::New<v8::Object>();
-            v8::Local<v8::Object> tilequery_properties_obj = Nan::New<v8::Object>();
             for (auto const& prop : feature->properties) {
-                properties_obj->Set(Nan::New<v8::String>(prop.first).ToLocalChecked(), get_property_value(prop.second));
+                set_property(prop, properties_obj);
```
Comment (mapsam) [inline src/vtquery.cpp:354]: @flippmoke this loop never ... _loops_. Am I referencing `feature->properties` correctly?

File: src/vtquery.cpp [hunk 2]
```diff
@@ -324,12 +342,12 @@ struct Worker : Nan::AsyncWorker {
...
             }
 
             // set properties.tilquery
+            v8::Local<v8::Object> tilequery_properties_obj = Nan::New<v8::Object>();
             tilequery_properties_obj->Set(Nan::New("distance").ToLocalChecked(), Nan::New<v8::Number>(feature->distance));
             std::string og_geom = getGeomTypeString(feature->original_geometry_type);
             tilequery_properties_obj->Set(Nan::New("geometry").ToLocalChecked(), Nan::New<v8::String>(og_geom).ToLocalChecked());
```
Comment (mapsam) [inline src/vtquery.cpp:361]: @flippmoke The original geometry type string shows up in the final GeoJSON, but `layer_name` doesn't (one line lower). Not sure why.

**Comment by mapsam** [comment]
This is currently blocked by https://github.com/mapbox/spatial-algorithms/pull/21

**Comment by mapsam** [comment]
Spatial algorithms ☝️  branch has been merged and has a dev release on Mason. Onward.

**Comment by mapsam** [comment]
Looks like -weffc++ errors are throwing:

```
../src/vtquery.cpp:266:49: error: comparing floating point with == or != is unsafe [-Werror=float-equal]
                         if (cp_info.distance == 0.0) {
                                                 ^~~
```

but it looks like this isn't a big deal according to [SO](https://stackoverflow.com/questions/11421756/weverything-yielding-comparing-floating-point-with-or-is-unsafe#11422298). I imagine I can bypass this by reversing the conditional to something like `if (cp_info.distance >= 0.0) { ... }`. Is this what you would do @flippmoke?

File: src/vtquery.cpp (hunk 2 above)
Comment (mapsam) [reply src/vtquery.cpp:361]: This has been fixed

File: src/vtquery.cpp (hunk 1 above)
Comment (mapsam) [reply src/vtquery.cpp:354]: Fixed!

**Comment by springmeyer** [comment]
@mapsam - awesome work on this. Did a quick review and here are my thoughts:

- do you have test coverage of sorted results? This PR looks like it changed the method of sorting and looks good to me - but do you have any tests that assert that the sorting of features is the same before and after? In particular do you have any tests that cover what happens when two or more features are the same exact distance? In this case what order do they end up in? Is it stable?
- I noticed you can remove some repeated code in the geometry handling. How about applying this?  https://gist.github.com/springmeyer/c3f88c31dc58e536f8b5dd5b8b015aaa
- I noticed (unrelated to this PR) you have a string called `uknown` - is that misspelling intended?
- The `Nan::New<v8::Array>` constructor takes a size, which will optimize its creation when you know the final size. In your case I think you do: you can pass `Nan::New<v8::Array>(results_queue_.size());` in one case (when creating the `features_array`) and `Nan::New<v8::Array>(2)` in another case (when creating coordinates). This will not likely improve performance in any measurable way but will nevertheless be more efficient, so it is worth applying I think.

**Review by springmeyer — APPROVED** [review]
Made minor comments above. This looks great otherwise, so approving.

**Comment by mapsam** [comment]
> do you have test coverage of sorted results? 

I don't - that feels like a really good test to add. I'll work on updating one of them to check if the values are the same in the resulting feature collection.

> I noticed you can remove some repeated code in the geometry handling. How about applying this? /springmeyer/c3f88c31dc58e536f8b5dd5b8b015aaa

Mind updating this link @springmeyer?

> string called uknown - is that misspelling intended?

Indeed it is. I will update it 👌 

My todos:

- [x] `Nan::New<v8::Array>` constructor size
- [x] update uknown spelling
- [x] implement @springmeyer's suggestion once I see it!
- [x] test: cover what happens when two or more features are the same exact distance
- [x] test: cover when results include a 0.0 point in polygon (ensure it is the first result)

**Comment by mapsam** [comment]
@springmeyer @flippmoke interesting quirk came up when testing if results come in the same order - it appears floating point numbers are different on mac versus linux. Do you suggest I round the numbers in the tests or should I round in the actual library? Example error (tests pass on mac locally for me)

```
not ok 174 same distance
  ---
    operator: equal
    expected: 30.320996160176545
    actual:   30.32099616026162
    stack: |-
      Error: same distance
```

**Comment by springmeyer** [comment]
> round in the actual library

👍  to handling in the library if possible. However, might be tricky to figure out where it is happening: is the difference arising from code in vtquery.cpp, or libvectortile, or spatial-algos, or boost::geometry?

**Comment by mapsam** [comment]
@springmeyer the distance values are call coming from the closest_point algorithm, so perhaps a combination of spatial-algorithms and boost::geometry. I think we'd need to round the number _after_ receiving it from the closest_point.

**Comment by springmeyer** [comment]
@mapsam - okay, if the difference is in spatial-algos, then I'm comfortable with you applying a fuzzy comparison in vtquery rather than exact. It looks like this is what spatial-algos is doing by using `BOOST_CHECK_CLOSE` with an epsilon: https://github.com/mapbox/spatial-algorithms/blob/master/test/unit/closest_point/test-closest-point.cpp

**Comment by mapsam** [comment]
@springmeyer how does this look as a viable checkClose function?

```javascript
function checkClose(a, b, epsilon) {
  return 1*(a-b) < epsilon;
}
```

And usage in tests with same epsilon as spatial-algorithms

```javsacript
assert.ok(checkClose(number_a, number_b, 1e-6));
```

**Comment by springmeyer** [comment]
looks great @mapsam!

**Comment by mapsam** [comment]
Great! Okay, [only one more thing I'm seeing](https://travis-ci.org/mapbox/vtquery/jobs/308726743#L925) - linux seems to return results with the same distance in a different order, despite the distance being the same number. Must they be in the same order? I would expect them to be in parsing order based on the operations of the priority queue, but this is hitting my expertise in how the priority queue resorts itself.

**Comment by springmeyer** [comment]
@mapsam - I'm not sure why, but I have a hunch and if I'm right @flippmoke can explain the details tomorrow. My hunch is that the sort is not "stable". In the past when using `std::sort` we've found we needed to move to http://en.cppreference.com/w/cpp/algorithm/stable_sort to get tests passing on osx and linux for fixtures. Since you are using priority_queue I wonder if https://lemire.me/blog/2017/03/13/stable-priority-queues/ is relevant. Again @flippmoke will know better than me, but this is a starting hunch.

**Comment by mapsam** [comment]
@springmeyer I lied so badly! 🙈 

We're never using distance from closest_point as a return, and are only using it in the conditional comparison. What we're returning is a value from cheap-ruler based on latitude and longitude inputs. So basically we convert vector tile coordinates into latitude and longitude coordinates ([cast to doubles](https://github.com/mapbox/vtquery/blob/7b79a90307d2ee2a049f340ef69fed8349563a0b/src/util.hpp#L92-L93)), then use cheap-ruler to calculate the distance between them in meters with [this function](https://github.com/mapbox/vtquery/blob/7b79a90307d2ee2a049f340ef69fed8349563a0b/src/util.hpp#L105-L110).

Numbers, as they move through vtquery (logged with std::clog):

```shell
# x,y
tilecoords:  601, 258
lnglat:      -87.7963, 41.8675
meters:      9.43636
ResObjDistance: 9.43636
js distance: 9.436356889343624
```

So, logging these numbers, it seems that closest_point and cheap-ruler work as expected and keep our numbers nice and rounded. It's when we pass from the v8 thread into a javascript value when we start getting sloppy. I think all javascript floating numbers are 64 bit, so not sure if there's anything we can do here other than continue asserting on the checkClose function?

**Comment by mapsam** [comment]
Just a confirmation that these tests are passing on macosx and not linux. Appears we cannot rely on order - @flippmoke have you seen priority_queue's react differently on different operating systems?

screen shot 2017-11-28 at 5 47 05 pm

**Comment by springmeyer** [comment]
@mapsam, nice debugging. As an extra assurance that the numbers are changing at the JS conversion I recommend adding a std::setprecision call to your std::clog call. Set the precision higher so that you can see more digits after the decimal (equal to the JS). What I’d look for is that even with higher display precision the ‘ResObjDistance’ output is still different.

**Comment by mapsam** [comment]
Nice @springmeyer - here's the output with std::setprecision(20):

```
tilecoords:     601, 258
lnglat:         -87.796286344528198242, 41.867499528053855329
meters:         9.4363568893436244878
ResObjDistance: 9.4363568893436244878
```

Not sure how to read this, though.

**Comment by springmeyer** [comment]
@mapsam - thanks for posting that. So that helps show that the precision output was misleading because it was rounding when the C++ number was printed. So:

```
ResObjDistance: 9.43636
```

Is actually:

```
ResObjDistance: 9.4363568893436244878
```

Which look identical when printed (other than the precision):

```C++
9.4363568893436244878 # C++
9.436356889343624 # JS
```

Am I following along right?

**Comment by mapsam** [comment]
@springmeyer yep, that's right. I think it'll be best to just make sure our tests are checking a rounded number, considering anything beyond 4 decimal points is in the micrometers of distance.

I tried to recreate the priority queue sorting differences in a docker container (ubuntu:16.04) without any luck. The sort results end up the same. Perhaps I'm not using the proper versions?

```
~/mug/priority_queue_diffs$ clang++ -c -std=c++14 main.cpp && clang++ -o main main.o && ./main
d_a: 13.7984
d_c: 13.7984
d_b: 4.5493
d_d: 1.3325
~/mug/priority_queue_diffs$ docker run -it pqd
root@a01a48cfd288:/usr/local/src# g++ -c -std=c++14 main.cpp && g++ -o main main.o && ./main
d_a: 13.7984
d_c: 13.7984
d_b: 4.5493
d_d: 1.3325
```

**Comment by springmeyer** [comment]
Sounds good on checking a less precise number overall in the tests @mapsam

As far as the priority queue issue which exact error is caused by that? Is it the error on travis where features properties are not matching the expected since they return in a different order on travis? If you could not replicate with a Dockerfile that is interesting. Could you post the Dockerfile?

**Comment by mapsam** [comment]
@springmeyer I've tried a few more attempts, and was able to recreate the error once, but I think I got lucky as I'm not able to get it compiled any longer. Now this branch includes a dockerfile you can build with `docker build -t vtquery .` and connect to it via `docker run -it vtquery` and then run `make test` inside, but I'm seeing some issues:

```
Error: /usr/local/src/lib/binding/vtquery.node: undefined symbol: __asan_report_store4
```

Not sure what's happening here, but perhaps I've set up the dockerfile incorrectly?

**Comment by springmeyer** [comment]
Remove the fsanitize=address from the Dockerfile and that error should go away. That worked for gzip-hpp but needs more tooling to work with a node-addon.

**Comment by mapsam** [comment]
Lovely, that did it @springmeyer. Thanks for the 👀! Now back to the investigation.

**Comment by mapsam** [comment]
I've removed the test for the time being and opened a ticket #44 to remind us of this in the future.

**Comment by mapsam** [comment]
[Failing tests](https://travis-ci.org/mapbox/vtquery/jobs/309709065) are due to travis regressions which have been ticketed in https://github.com/mapbox/node-cpp-skel/issues/93

**Comment by mapsam** [comment]
Thanks @flippmoke and @springmeyer for the help here!

**Comment by springmeyer** [comment]
epic PR @mapsam 🙌

**Pull request merged**